dev.mk.in

```diff
@@ -1,245 +0,0 @@
-    $(non_third_party_sources) \
-    $(third_party_sources) \
-    $(headers) \
-    $(test_sources) \
-    CONTRIBUTING.md \
-    GPL-3.0.txt \
-    LICENSE.adoc \
-    Makefile.in \
-    README.md \
-    autogen.sh \
-    config.guess \
-    config.h.in \
-    config.sub \
-    configure \
-    configure.ac \
-    dev.mk.in \
-    doc/AUTHORS.adoc \
-    doc/MANUAL.adoc \
-    doc/NEWS.adoc \
-    install-sh \
-    m4 \
-    src/main.cpp \
-    src/third_party/minitrace.c \
-    test/run \
-    test/suites/*.bash \
-    unittest/catch2_tests.cpp \
-    unittest/catch2_tests.hpp
-
-dist_files = \
-    $(addprefix $(srcdir)/, $(source_dist_files)) \
-    $(built_dist_files)
-
-ifneq ($(shell sed 's/.*"\(.*\)".*/\1/' src/version.cpp 2>/dev/null),$(version))
-  $(shell echo 'extern const char CCACHE_VERSION[]; const char CCACHE_VERSION[] = "$(version)";' >src/version.cpp)
-endif
-src/version.o: src/version.cpp
-
-.PHONY: dist
-dist: $(dist_archives)
-
-$(dist_archives): $(dist_files)
-	tmpdir=$$(mktemp -d /tmp/tmp-ccache-dist.XXXXXX) && \
-	dir=$$tmpdir/$(dist_dir) && \
-	mkdir $$dir && \
-	(cd $(srcdir) && \
-	 rsync -r --relative $(source_dist_files) $$dir) && \
-	cp $(srcdir)/doc/INSTALL-from-release-archive.md $$dir/INSTALL.md && \
-	(cd $(builddir) && \
-	 rsync -r --relative $(built_dist_files) $$dir) && \
-	echo "Remove this file to enable developer mode." >$$dir/dev_mode_disabled && \
-	(cd $$tmpdir && \
-	 tarcompression= && \
-	 case $@ in \
-	     *.gz) tarcompression=-z ;; \
-	     *.xz) tarcompression=-J ;; \
-	 esac && \
-	 tar -c $$tarcompression -f $(CURDIR)/$@ $(dist_dir)) && \
-	rm -rf $$tmpdir
-
-# $(1): extra configure options
-define do_distcheck
-	tmpdir=$$(mktemp -d /tmp/tmp-ccache-distcheck.XXXXXX) && \
-	(cd $$tmpdir && \
-	 tar xf $(CURDIR)/$< && \
-	 mkdir -p $(dist_dir)/build && \
-	 chmod -R a-w $(dist_dir) && \
-	 chmod u+w $(dist_dir)/build && \
-	 cd $(dist_dir)/build && \
-	 ../configure --enable-more-warnings --prefix=$$tmpdir/root --sysconfdir=/etc $(1) && \
-	 $(MAKE) install V=1 && \
-	 $(MAKE) installcheck V=1) && \
-	chmod -R u+w $$tmpdir/$(dist_dir) && \
-	rm -rf $$tmpdir
-endef
-
-.PHONY: distcheck
-distcheck: $(firstword $(dist_archives))
-	$(call do_distcheck, CFLAGS=-Werror CXXFLAGS=-Werror)
-	$(call do_distcheck, --with-libzstd-from-internet --with-libb2-from-internet)
-	$(call do_distcheck, CC=clang CXX=clang++)
-
-.PHONY: docs
-docs: $(generated_docs)
-
-%.html: %.adoc
-	@mkdir -p $(@D)
-	$(if $(quiet),@echo "  ASCIIDOC $@")
-	$(Q)$(ASCIIDOC) -o $@ -a revnumber=$(version) -a toc -b xhtml11 $<
-
-%.xml: %.adoc
-	@mkdir -p $(@D)
-# Make literals stand out as bold in the man page:
-	$(if $(quiet),@echo "  ASCIIDOC $@")
-	$(Q)$(ASCIIDOC) -a revnumber=$(version) -d manpage -b docbook -o - $< | \
-	  perl -pe 's!<literal>(.*?)</literal>!<emphasis role="strong">\1</emphasis>!g' >$@
-
-doc/ccache.1: doc/MANUAL.xml
-	$(if $(quiet),@echo "  A2X      $@")
-	$(Q)$(A2X) --doctype manpage --format manpage $<
```

Comment:
Could anyone explain what is happening here?
I don't understand what is actually executed between all this make magic.

Reply:
Are you talking about the [pattern rules](https://www.gnu.org/software/make/manual/html_node/Pattern-Rules.html) syntax/semantics or the commands they execute (the "recipes" in Make terminology)? What do you think is magic about them?